README.md
# pushbullet-automation-haskell

A Haskell re-implementation of [Pushbullet Automation.](https://github.com/timTam97/pushbullet-automation)


ChangeLog.md
# Changelog for pb-automation-haskell

## Unreleased changes


src/Test.hs
{-# LANGUAGE OverloadedStrings #-}

module Test where

import Control.Monad.IO.Class ()
import Data.Aeson
  ( FromJSON (parseJSON),
    Value (Object),
    decode,
    (.:),
  )
import Data.Aeson.TH ()
import Data.Aeson.Types (prependFailure, typeMismatch)
import qualified Data.ByteString.Lazy.Char8 as BL
import Data.IORef (newIORef, readIORef, writeIORef)
import Data.Maybe ()
import Network.HTTP.Req ()
import System.Environment

data Push = Push {pushes :: [SubPushes]} deriving (Show)

data SubPushes = SubPushes
  { title :: String,
    body :: String,
    modified :: Double,
    dismissed :: Bool,
    sender_name :: String
  }
  deriving (Show)

instance FromJSON Push where
  parseJSON (Object v) =
    Push
      <$> v .: "pushes"
  parseJSON invalid =
    prependFailure
      "parsing Coord failed, "
      (typeMismatch "Object" invalid)

instance FromJSON SubPushes where
  parseJSON (Object v) =
    SubPushes
      <$> v .: "title"
      <*> v .: "body"
      <*> v .: "modified"
      <*> v .: "dismissed"
      <*> v .: "sender_name"

getFile :: IO String
getFile = readFile "src\\test.txt"

test :: IO ()
test = do
  str <- getFile
  let ok = decode $ BL.pack str :: Maybe Push
  case ok of
    Just a -> print a
    Nothing -> print "bruh"

test1 :: IO ()
test1 = do
  pot <- newIORef 2
  -- val <- pot
  hmm <- readIORef pot
  case hmm of
    2 -> writeIORef pot 50
  val <- readIORef pot
  print val

test2 :: IO ()
test2 = do
  tok <- getEnv "PB_TOKEN_USR"
  print tok


src/Sys.hs
{-# LANGUAGE ForeignFunctionInterface #-}

module Sys (setSuspendState, lockWorkStation, shutdownSystem, getComputerName) where

import Control.Exception (SomeException, try)
import Foreign.Ptr (FunPtr, castPtrToFunPtr)
import System.Process (callCommand)
import System.Win32.DLL (getProcAddress, loadLibraryEx)
import System.Win32.Info.Computer (getComputerName)
import System.Win32.Types (Addr, nullHANDLE)

type TripleBool = Bool -> Bool -> Bool -> IO ()

foreign import ccall "dynamic"
  setSuspendState0 ::
    FunPtr TripleBool ->
    TripleBool

foreign import ccall "dynamic" lockWorkStation0 :: FunPtr (IO ()) -> IO ()

getFuncAddr :: String -> String -> IO Addr
getFuncAddr dllName methodName = do
  dll <- loadLibraryEx dllName nullHANDLE 0x00000800
  getProcAddress dll methodName

-- true: hibernate
-- false: sleep
setSuspendState :: Bool -> IO ()
setSuspendState b1 = do
  addr <- getFuncAddr "PowrProf.dll" "SetSuspendState"
  setSuspendState0 (castPtrToFunPtr addr) b1 False False

lockWorkStation :: IO ()
lockWorkStation = do
  addr <- getFuncAddr "User32.dll" "LockWorkStation"
  lockWorkStation0 $ castPtrToFunPtr addr

shutdownSystem :: IO ()
shutdownSystem = do
  try (callCommand "shutdown /s /t 0") :: IO (Either SomeException ())
  pure ()


src/Web.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TemplateHaskell #-}

module Web (mainLoop) where

import Control.Exception (try)
import Control.Monad (forever, when)
import Control.Monad.Loops (whileJust_)
import Data.Aeson
  ( Options (fieldLabelModifier),
    decode,
    defaultOptions,
  )
import Data.Aeson.TH (deriveJSON)
import qualified Data.ByteString.Char8 as BS
import qualified Data.ByteString.Lazy.Char8 as BL
import Data.IORef (IORef, newIORef, readIORef, writeIORef)
import Data.Maybe (fromJust, isJust)
import Data.Text (Text, unpack)
import Network.HTTP.Req
  ( GET (GET),
    NoReqBody (NoReqBody),
    defaultHttpConfig,
    header,
    https,
    jsonResponse,
    req,
    responseBody,
    runReq,
    (/:),
    (=:),
  )
import qualified Network.WebSockets as WS
import System.Timeout (timeout)
import Util
  ( currentUnixTime,
    fromJSONValue,
    getToken,
    processCommand,
    writeLog,
    getComputerName
  )
import qualified Wuss as WWS

data RecData = RecData {mainType :: String, subtype :: Maybe String}
  deriving (Show)

newtype Push = Push {pushes :: [SubPushes]}
  deriving (Show)

data SubPushes = SubPushes
  { title :: String,
    body :: String,
    modified :: Double,
    dismissed :: Bool,
    sender_name :: String
  }
  deriving (Show)

-- https://stackoverflow.com/questions/48474587/how-to-deal-with-haskells-reserved-keywords-in-record-fields
$( deriveJSON
     defaultOptions
       { fieldLabelModifier =
           \x -> if x == "mainType" then "type" else x
       }
     ''RecData
 )
$(deriveJSON defaultOptions ''Push)
$(deriveJSON defaultOptions ''SubPushes)

interrogateResponse :: Push -> Integer -> IO Integer
interrogateResponse p n = do
  computerName <- getComputerName
  if title subPush == computerName
    && body subPush /= ""
    && not (dismissed subPush)
    then do
      processCommand $ body subPush
      pure $ floor $ modified subPush
    else pure n
  where
    subPush = head $ pushes p

grabPush :: Integer -> IO (Maybe Push)
grabPush t = do
  token <- getToken
  runReq defaultHttpConfig $ do
    r <-
      req
        GET
        (https "api.pushbullet.com" /: "v2" /: "pushes")
        NoReqBody
        jsonResponse
        ( header "Access-Token" (BS.pack token)
            <> ("modified_after" =: (show t :: String))
            <> ("active" =: ("true" :: String))
            <> ("limit" =: (1 :: Integer))
        )
    -- https://williamyaoh.com/posts/2019-10-19-a-cheatsheet-to-json-handling.html
    pure $ fromJSONValue (responseBody r)

analyzePing :: IORef Integer -> Text -> IO ()
analyzePing tme msg = do
  let received = decode $ BL.pack $ unpack msg :: Maybe RecData
  case received of
    Just a ->
      when (mainType a == "tickle") $ do
        numTime <- readIORef tme
        push <- grabPush numTime
        when (isJust push) $ do
          newTime <- interrogateResponse (fromJust push) numTime
          writeIORef tme newTime

app :: WS.ClientApp ()
app conn = do
  time <- currentUnixTime
  recentTime <- newIORef time
  writeLog "Connected"
  whileJust_
    (timeout 35000000 $ do WS.receiveData conn :: IO Text)
    (analyzePing recentTime)
  writeLog "Exiting"
  WS.sendClose conn ("Bye!" :: Text)

mainLoop :: IO ()
mainLoop = do
  token <- getToken
  forever $ do
    exc <-
      try $
        WWS.runSecureClient
          "stream.pushbullet.com"
          443
          ("/websocket/" <> token)
          app ::
        IO (Either WS.ConnectionException ())
    case exc of
      Left a -> writeLog $ "Caught exception " ++ show a ++ ", continuing"
      Right _ -> writeLog "Restarting connection"


src/Util.hs
module Util
  ( writeLog,
    currentUnixTime,
    getToken,
    fromJSONValue,
    processCommand,
    reformatFile,
    getComputerName
  )
where

import Data.Aeson.Types (FromJSON (parseJSON), Value, parseMaybe)
import Data.Text (unpack)
import Data.Time (defaultTimeLocale, formatTime, getZonedTime)
import Data.Time.Clock.POSIX (getPOSIXTime)
import Ormolu (defaultConfig, ormoluFile)
import qualified Sys as SYS
import System.Environment (getEnv)
import System.IO (IOMode (AppendMode), hClose, hPutStrLn, openFile)

getComputerName :: IO String
getComputerName = SYS.getComputerName

reformatFile :: String -> IO ()
reformatFile fileName = do
  txt <- ormoluFile defaultConfig fileName
  writeFile fileName $ unpack txt

logTimeStr :: IO String
logTimeStr = do
  now <- getZonedTime
  let str = formatTime defaultTimeLocale "%Y-%m-%d %H:%M:%S" now
  pure $ "[" <> str <> "] "

writeLog :: String -> IO ()
writeLog str = do
  file <- openFile ".\\run.log" AppendMode
  time <- logTimeStr
  hPutStrLn file $ time <> str
  hClose file

currentUnixTime :: IO Integer
currentUnixTime = floor <$> getPOSIXTime

getToken :: IO String
getToken = getEnv "PB_TOKEN_SYS"

fromJSONValue :: FromJSON a => Value -> Maybe a
fromJSONValue = parseMaybe parseJSON

processCommand :: String -> IO ()
processCommand c =
  writeLog ("Command: " <> c)
    >> case c of
      "sleep" -> SYS.setSuspendState False
      "hibernate" -> SYS.setSuspendState True
      "shut down" -> SYS.shutdownSystem
      "lock" -> SYS.lockWorkStation


src/Lib.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TemplateHaskell #-}

module Lib where

-- import Data.List
-- import qualified Network.WebSockets.Client as WS

-- someFunc :: IO ()
-- someFunc = putStrLn "someFunc"
import Data.IORef ( newIORef, readIORef, writeIORef )
import Control.Monad (forever, when)
import Control.Monad.IO.Class ()
import Control.Monad.Trans (liftIO)
import Data.Aeson
    ( decode,
      defaultOptions,
      FromJSON(parseJSON),
      Options(fieldLabelModifier),
      Value )
-- import Data.Aeson (decode, defaultOptions)
import Data.Aeson.TH ( deriveJSON )
import Data.Aeson.Types ( parseMaybe )
import qualified Data.ByteString.Char8 as BS
import qualified Data.ByteString.Lazy.Char8 as BL
import Data.Text (Text, unpack)
import qualified Data.Text.IO as T
import Data.Time.Clock.POSIX ( getPOSIXTime )
import Network.HTTP.Req
    ( (/:),
      (=:),
      defaultHttpConfig,
      header,
      https,
      jsonResponse,
      req,
      responseBody,
      runReq,
      GET(GET),
      NoReqBody(NoReqBody) )
import qualified Network.WebSockets as WS
import System.IO (IOMode (ReadMode), hGetLine, openFile)
import qualified Wuss as WWS
import Data.Maybe ( fromJust, isJust )
import qualified Sys as SYS
import Debug.Trace ()
data RecData = RecData {mainType :: String, subtype :: Maybe String}
  deriving (Show)

newtype Push
  = Push {pushes :: [SubPushes]}
  deriving Show

data SubPushes = SubPushes {
  title :: String,
  body :: String,
  modified :: Double,
  dismissed :: Bool,
  sender_name :: String
} deriving (Show)

-- https://stackoverflow.com/questions/48474587/how-to-deal-with-haskells-reserved-keywords-in-record-fields
$( deriveJSON
     defaultOptions
       { fieldLabelModifier =
           \x -> if x == "mainType" then "type" else x
       }
     ''RecData
 )
$(deriveJSON defaultOptions ''Push)
$(deriveJSON defaultOptions ''SubPushes)

currentUnixTime :: IO Integer
currentUnixTime = do floor <$> getPOSIXTime

getToken :: IO String
getToken = do
  handle <- openFile "src\\token.txt" ReadMode
  hGetLine handle

fromJSONValue :: FromJSON a => Value -> Maybe a
fromJSONValue = parseMaybe parseJSON

processCommand :: String -> IO ()
processCommand c = 
  case c of
    "sleep" -> SYS.setSuspendState False
    "hibernate" -> SYS.setSuspendState True
    "shut down" -> undefined
    "lock" -> SYS.lockWorkStation

interrogateResponse :: Push -> Integer -> IO Integer
interrogateResponse p n =
  if title subPush == "LENOVO-TIM" 
    && body subPush /= ""
    && not (dismissed subPush)
  then do
    processCommand $ body subPush
    pure $ floor $ modified subPush
  else
    pure n
  where
    subPush = head $ pushes p

grabPush :: Integer -> IO (Maybe Push)
grabPush t = do
  token <- getToken
  runReq defaultHttpConfig $ do
    r <-
      req
        GET
        (https "api.pushbullet.com" /: "v2" /: "pushes")
        NoReqBody
        jsonResponse
        ( header "Access-Token" (BS.pack token)
            <> ("modified_after" =: (show t :: String))
            <> ("active" =: ("true" :: String))
            <> ("limit" =: (1 :: Integer))
        )
    -- https://williamyaoh.com/posts/2019-10-19-a-cheatsheet-to-json-handling.html
    let parsed = fromJSONValue (responseBody r) :: Maybe Push
    liftIO $ print parsed
    pure parsed
    -- liftIO $ print parsed

app :: WS.ClientApp ()
app conn = do
  time <- currentUnixTime
  recentTime <- newIORef time
  putStrLn "Connected!"
  forever $ do
    msg <- WS.receiveData conn
    let received = decode $ BL.pack $ unpack msg :: Maybe RecData
    case received of
      Just a ->
        when (mainType a == "tickle") $ do
          numTime <- readIORef recentTime
          push <- grabPush numTime
          when (isJust push) (do
          newTime <- interrogateResponse (fromJust push) numTime
          writeIORef recentTime newTime)
    print received
    liftIO $ T.putStrLn msg

  WS.sendClose conn ("Bye!" :: Text)

mainLoop :: IO ()
mainLoop = do
  token <- getToken
  WWS.runSecureClient "stream.pushbullet.com" 443 ("/websocket/" <> token) app
